File: src/api/endpoints/runners/lambdas/common/runner_labels.py
"""Runner label parsing and validation."""

import json
import os
import re
from pathlib import Path
from typing import Any


class LabelParseError(Exception):
    """Error parsing runner labels."""


class LabelValidationError(Exception):
    """Error validating runner labels."""


def _find_config_file() -> Path:
    """Find the runners.json configuration file."""
    # Try ETC_PATH environment variable
    etc_path = os.environ.get("ETC_PATH")
    if etc_path:
        config_path = Path(etc_path) / "runners.json"
        if config_path.exists():
            return config_path

    # Try relative to this file (Lambda deployment: bundled in common/etc/)
    layer_path = Path(__file__).parent / "etc" / "runners.json"
    if layer_path.exists():
        return layer_path

    # Try repo root structure (local development)
    repo_path = Path(__file__).parent.parent.parent.parent.parent.parent.parent
    repo_config = repo_path / "etc" / "runners.json"
    if repo_config.exists():
        return repo_config

    # Try current working directory
    cwd_path = Path.cwd() / "etc" / "runners.json"
    if cwd_path.exists():
        return cwd_path

    raise FileNotFoundError(
        "Could not find etc/runners.json. "
        "Set ETC_PATH environment variable or run from repo root."
    )


def _load_config() -> dict[str, Any]:
    """Load configuration from JSON file."""
    config_path = _find_config_file()
    with open(config_path, encoding="utf-8") as f:
        return json.load(f)


# Load configuration from JSON (single source of truth)
_config = _load_config()
_labels = _config.get("labels", {})

# Valid label values from JSON
PLATFORMS: set[str] = set(_labels.get("platforms", ["ecs", "ec2"]))
PRICING_MODELS: set[str] = set(_labels.get("pricing_models", ["spot", "on-demand"]))

# ECS labels from JSON
_ecs_config = _labels.get("ecs", {})
ECS_COMPUTE: set[str] = set(_ecs_config.get("compute_types", ["fargate"]))
ECS_ARCHITECTURES: set[str] = set(_ecs_config.get("architectures", ["x86", "arm"]))

# EC2 labels from JSON
_ec2_config = _labels.get("ec2", {})
EC2_COMPUTE: set[str] = set(_ec2_config.get("compute_types", []))
EC2_ARCHITECTURES: set[str] = set(_ec2_config.get("architectures", []))
_EC2_COMPUTE_REQUIRES_ARCH: set[str] = set(
    _ec2_config.get("compute_requires_arch", [])
)
_EC2_COMPUTE_FORBIDS_ARCH: set[str] = set(_ec2_config.get("compute_forbids_arch", []))

# All compute types (derived)
COMPUTE_TYPES: set[str] = ECS_COMPUTE | EC2_COMPUTE

# All architecture labels
ALL_ARCHITECTURES: set[str] = ECS_ARCHITECTURES | EC2_ARCHITECTURES

# Instance type mapping from JSON
_instance_map = _labels.get("instance_type_map", {})
_INSTANCE_TYPE_MAP: dict[str, str] = {}
for _compute, _arch_map in _instance_map.items():
    if isinstance(_arch_map, dict):
        for _arch, _instance_type in _arch_map.items():
            _INSTANCE_TYPE_MAP[f"{_compute}:{_arch}"] = _instance_type
    else:
        # No architecture (gpu, fpga)
        _INSTANCE_TYPE_MAP[f"{_compute}:None"] = _arch_map

# ECS task architecture mapping from JSON
_ECS_TASK_ARCHITECTURE_MAP: dict[str, str] = _labels.get(
    "ecs_task_architecture_map", {}
)

# ECS Fargate configuration from JSON
_fargate_config = _config.get("fargate", {})
_ECS_FARGATE_CONFIG = {
    "cpu": _fargate_config.get("cpu", "4096"),
    "memory": _fargate_config.get("memory", "16384"),
}

# Runner ID pattern
_RUNNER_ID_PATTERN = re.compile(r"^runner-(\d+)$")


def parse_labels(label_list: list[str]) -> dict[str, Any]:
    """Parse job labels to extract platform, compute, architecture, pricing, runner ID.

    Args:
        label_list: List of labels from the workflow job

    Returns:
        Dictionary with platform, compute, pricing, runnerId, architecture

    Raises:
        LabelParseError: If required labels are missing or duplicated
    """
    if not label_list:
        raise LabelParseError("Label list cannot be empty")

    label_set = set(label_list)

    # Extract platform
    platform_labels = [label for label in label_set if label in PLATFORMS]
    if not platform_labels:
        raise LabelParseError(
            f"Missing platform label. Must be one of: {', '.join(sorted(PLATFORMS))}"
        )
    if len(platform_labels) > 1:
        raise LabelParseError(
            f"Multiple platform labels found: {', '.join(sorted(platform_labels))}"
        )
    platform = platform_labels[0]

    # Extract compute type
    compute_labels = [label for label in label_set if label in COMPUTE_TYPES]
    if not compute_labels:
        raise LabelParseError(
            f"Missing compute label. Must be one of: "
            f"{', '.join(sorted(COMPUTE_TYPES))}"
        )
    if len(compute_labels) > 1:
        raise LabelParseError(
            f"Multiple compute labels found: {', '.join(sorted(compute_labels))}"
        )
    compute = compute_labels[0]

    # Extract architecture (may be None for gpu/fpga)
    arch_labels = [label for label in label_set if label in ALL_ARCHITECTURES]
    if len(arch_labels) > 1:
        raise LabelParseError(
            f"Multiple architecture labels found: {', '.join(sorted(arch_labels))}"
        )
    architecture = arch_labels[0] if arch_labels else None

    # Extract pricing model
    pricing_labels = [label for label in label_set if label in PRICING_MODELS]
    if not pricing_labels:
        raise LabelParseError(
            f"Missing pricing label. Must be one of: "
            f"{', '.join(sorted(PRICING_MODELS))}"
        )
    if len(pricing_labels) > 1:
        raise LabelParseError(
            f"Multiple pricing labels found: {', '.join(sorted(pricing_labels))}"
        )
    pricing = pricing_labels[0]

    # Extract runner ID
    runner_id = None
    for label in label_list:
        if _RUNNER_ID_PATTERN.match(label):
            runner_id = label
            break

    if not runner_id:
        raise LabelParseError(
            "Missing runner ID label. Must match pattern: runner-{number}"
        )

    return {
        "platform": platform,
        "compute": compute,
        "pricing": pricing,
        "runnerId": runner_id,
        "architecture": architecture,
    }


def validate_labels(parsed: dict[str, Any]) -> None:
    """Validate parsed labels for platform-specific requirements.

    Args:
        parsed: Parsed labels from parse_labels()

    Raises:
        LabelValidationError: If labels are invalid for the platform
    """
    platform = parsed["platform"]
    compute = parsed["compute"]
    architecture = parsed["architecture"]

    # ECS validation
    if platform == "ecs":
        if compute not in ECS_COMPUTE:
            raise LabelValidationError(
                f"ECS platform only supports compute types: "
                f"{', '.join(sorted(ECS_COMPUTE))}. Got: {compute}"
            )
        # ECS requires architecture
        if architecture is None:
            raise LabelValidationError(
                f"ECS platform requires architecture label. "
                f"Must be one of: {', '.join(sorted(ECS_ARCHITECTURES))}"
            )
        # ECS architecture must be valid
        if architecture not in ECS_ARCHITECTURES:
            raise LabelValidationError(
                f"ECS platform only supports architectures: "
                f"{', '.join(sorted(ECS_ARCHITECTURES))}. Got: {architecture}"
            )

    # EC2 validation
    if platform == "ec2":
        if compute not in EC2_COMPUTE:
            raise LabelValidationError(
                f"EC2 platform only supports compute types: "
                f"{', '.join(sorted(EC2_COMPUTE))}. Got: {compute}"
            )

        # Check architecture requirements based on compute type
        if compute in _EC2_COMPUTE_REQUIRES_ARCH:
            if architecture is None:
                raise LabelValidationError(
                    f"EC2 compute type '{compute}' requires architecture label. "
                    f"Must be one of: {', '.join(sorted(EC2_ARCHITECTURES))}"
                )
            if architecture not in EC2_ARCHITECTURES:
                raise LabelValidationError(
                    f"EC2 platform only supports architectures: "
                    f"{', '.join(sorted(EC2_ARCHITECTURES))}. Got: {architecture}"
                )

        if compute in _EC2_COMPUTE_FORBIDS_ARCH:
            if architecture is not None:
                raise LabelValidationError(
                    f"EC2 compute type '{compute}' does not support "
                    f"architecture label. Got: {architecture}"
                )


def get_instance_type(parsed: dict[str, Any]) -> str | None:
    """Get EC2 instance type for the parsed labels.

    Args:
        parsed: Parsed labels from parse_labels()

    Returns:
        Instance type string or None for ECS
    """
    if parsed["platform"] == "ecs":
        return None
    key = f"{parsed['compute']}:{parsed['architecture']}"
    return _INSTANCE_TYPE_MAP.get(key)


def get_ecs_config(parsed: dict[str, Any]) -> dict[str, str] | None:
    """Get ECS Fargate configuration for the parsed labels.

    Args:
        parsed: Parsed labels from parse_labels()

    Returns:
        Dictionary with cpu and memory, or None for non-ECS/non-Fargate
    """
    if parsed["platform"] != "ecs":
        return None
    if parsed["compute"] == "fargate":
        return dict(_ECS_FARGATE_CONFIG)
    return None


def get_task_architecture(parsed: dict[str, Any]) -> str | None:
    """Get ECS task architecture for the parsed labels.

    Args:
        parsed: Parsed labels from parse_labels()

    Returns:
        ECS task architecture (X86_64 or ARM64) or None
    """
    if parsed["platform"] != "ecs":
        return None
    if parsed["architecture"] is None:
        return None
    return _ECS_TASK_ARCHITECTURE_MAP.get(parsed["architecture"])


def is_spot(parsed: dict[str, Any]) -> bool:
    """Check if the parsed labels indicate spot pricing.

    Args:
        parsed: Parsed labels from parse_labels()

    Returns:
        True if spot pricing, False otherwise
    """
    return parsed["pricing"] == "spot"


def get_runner_id_number(parsed: dict[str, Any]) -> int:
    """Extract the numeric part of the runner ID.

    Args:
        parsed: Parsed labels from parse_labels()

    Returns:
        Runner ID number

    Raises:
        LabelParseError: If runner ID format is invalid
    """
    match = _RUNNER_ID_PATTERN.match(parsed["runnerId"])
    if match:
        return int(match.group(1))
    raise LabelParseError(f"Invalid runner ID format: {parsed['runnerId']}")

File: src/api/endpoints/runners/lambdas/layers/common/runner_labels.js
const fs = require('fs');
const path = require('path');

class LabelParseError extends Error {
  constructor(message) {
    super(message);
    this.name = 'LabelParseError';
  }
}

class LabelValidationError extends Error {
  constructor(message) {
    super(message);
    this.name = 'LabelValidationError';
  }
}

function findConfigFile() {
  // Try ETC_PATH environment variable
  const etcPath = process.env.ETC_PATH;
  if (etcPath) {
    const configPath = path.join(etcPath, 'runners.json');
    if (fs.existsSync(configPath)) {
      return configPath;
    }
  }

  // Try relative to this file (Lambda layer deployment: /opt/nodejs/runners-layer/etc/)
  const layerPath = path.join(__dirname, 'etc', 'runners.json');
  if (fs.existsSync(layerPath)) {
    return layerPath;
  }

  // Try repo root structure (local development: src/api/endpoints/runners/lambdas/layer/)
  const repoPath = path.join(__dirname, '..', '..', '..', '..', '..', '..', 'etc', 'runners.json');
  if (fs.existsSync(repoPath)) {
    return repoPath;
  }

  // Try current working directory
  const cwdPath = path.join(process.cwd(), 'etc', 'runners.json');
  if (fs.existsSync(cwdPath)) {
    return cwdPath;
  }

  throw new Error(
    'Could not find etc/runners.json. ' +
    'Set ETC_PATH environment variable or run from repo root.'
  );
}

function loadConfig() {
  const configPath = findConfigFile();
  const content = fs.readFileSync(configPath, 'utf-8');
  return JSON.parse(content);
}

// Load configuration from JSON (single source of truth)
const config = loadConfig();
const labels = config.labels || {};

// Valid label values from JSON
const PLATFORMS = new Set(labels.platforms || ['ecs', 'ec2']);
const PRICING_MODELS = new Set(labels.pricing_models || ['spot', 'on-demand']);

// ECS labels from JSON
const ecsConfig = labels.ecs || {};
const ECS_COMPUTE = new Set(ecsConfig.compute_types || ['fargate']);
const ECS_ARCHITECTURES = new Set(ecsConfig.architectures || ['x86', 'arm']);

// EC2 labels from JSON
const ec2Config = labels.ec2 || {};
const EC2_COMPUTE = new Set(ec2Config.compute_types || []);
const EC2_ARCHITECTURES = new Set(ec2Config.architectures || []);
const EC2_COMPUTE_REQUIRES_ARCH = new Set(ec2Config.compute_requires_arch || []);
const EC2_COMPUTE_FORBIDS_ARCH = new Set(ec2Config.compute_forbids_arch || []);

// All compute types (derived)
const COMPUTE_TYPES = new Set([...ECS_COMPUTE, ...EC2_COMPUTE]);

// All architecture labels
const ALL_ARCHITECTURES = new Set([...ECS_ARCHITECTURES, ...EC2_ARCHITECTURES]);

// Instance type mapping from JSON
const instanceMap = labels.instance_type_map || {};
const INSTANCE_TYPE_MAP = {};
for (const [compute, archMap] of Object.entries(instanceMap)) {
  if (typeof archMap === 'object' && archMap !== null) {
    for (const [arch, instanceType] of Object.entries(archMap)) {
      INSTANCE_TYPE_MAP[`${compute}:${arch}`] = instanceType;
    }
  } else {
    // No architecture (gpu, fpga)
    INSTANCE_TYPE_MAP[`${compute}:null`] = archMap;
  }
}

// ECS task architecture mapping from JSON
const ECS_TASK_ARCHITECTURE_MAP = labels.ecs_task_architecture_map || {};

// ECS Fargate configuration from JSON
const fargateConfig = config.fargate || {};
const ECS_FARGATE_CONFIG = {
  cpu: fargateConfig.cpu || '4096',
  memory: fargateConfig.memory || '16384'
};

// Runner ID pattern
const RUNNER_ID_PATTERN = /^runner-(\d+)$/;

function parseLabels(labelList) {
  if (!labelList || labelList.length === 0) {
    throw new LabelParseError('Label list cannot be empty');
  }

  const labelSet = new Set(labelList);

  // Extract platform
  const platformLabels = [...labelSet].filter(l => PLATFORMS.has(l));
  if (platformLabels.length === 0) {
    throw new LabelParseError(
      `Missing platform label. Must be one of: ${[...PLATFORMS].sort().join(', ')}`
    );
  }
  if (platformLabels.length > 1) {
    throw new LabelParseError(
      `Multiple platform labels found: ${platformLabels.sort().join(', ')}`
    );
  }
  const platform = platformLabels[0];

  // Extract compute type
  const computeLabels = [...labelSet].filter(l => COMPUTE_TYPES.has(l));
  if (computeLabels.length === 0) {
    throw new LabelParseError(
      `Missing compute label. Must be one of: ${[...COMPUTE_TYPES].sort().join(', ')}`
    );
  }
  if (computeLabels.length > 1) {
    throw new LabelParseError(
      `Multiple compute labels found: ${computeLabels.sort().join(', ')}`
    );
  }
  const compute = computeLabels[0];

  // Extract architecture (may be null for gpu/fpga)
  const archLabels = [...labelSet].filter(l => ALL_ARCHITECTURES.has(l));
  if (archLabels.length > 1) {
    throw new LabelParseError(
      `Multiple architecture labels found: ${archLabels.sort().join(', ')}`
    );
  }
  const architecture = archLabels.length > 0 ? archLabels[0] : null;

  // Extract pricing model
  const pricingLabels = [...labelSet].filter(l => PRICING_MODELS.has(l));
  if (pricingLabels.length === 0) {
    throw new LabelParseError(
      `Missing pricing label. Must be one of: ${[...PRICING_MODELS].sort().join(', ')}`
    );
  }
  if (pricingLabels.length > 1) {
    throw new LabelParseError(
      `Multiple pricing labels found: ${pricingLabels.sort().join(', ')}`
    );
  }
  const pricing = pricingLabels[0];

  // Extract runner ID
  let runnerId = null;
  for (const label of labelList) {
    const match = label.match(RUNNER_ID_PATTERN);
    if (match) {
      runnerId = label;
      break;
    }
  }

  if (!runnerId) {
    throw new LabelParseError(
      'Missing runner ID label. Must match pattern: runner-{number}'
    );
  }

  return {
    platform,
    compute,
    pricing,
    runnerId,
    architecture
  };
}

function validateLabels(parsed) {
  // ECS validation
  if (parsed.platform === 'ecs') {
    if (!ECS_COMPUTE.has(parsed.compute)) {
      throw new LabelValidationError(
        `ECS platform only supports compute types: ${[...ECS_COMPUTE].sort().join(', ')}. ` +
        `Got: ${parsed.compute}`
      );
    }
    // ECS requires architecture
    if (parsed.architecture === null) {
      throw new LabelValidationError(
        `ECS platform requires architecture label. ` +
        `Must be one of: ${[...ECS_ARCHITECTURES].sort().join(', ')}`
      );
    }
    // ECS architecture must be valid
    if (!ECS_ARCHITECTURES.has(parsed.architecture)) {
      throw new LabelValidationError(
        `ECS platform only supports architectures: ${[...ECS_ARCHITECTURES].sort().join(', ')}. ` +
        `Got: ${parsed.architecture}`
      );
    }
  }

  // EC2 validation
  if (parsed.platform === 'ec2') {
    if (!EC2_COMPUTE.has(parsed.compute)) {
      throw new LabelValidationError(
        `EC2 platform only supports compute types: ${[...EC2_COMPUTE].sort().join(', ')}. ` +
        `Got: ${parsed.compute}`
      );
    }

    // Check architecture requirements based on compute type
    if (EC2_COMPUTE_REQUIRES_ARCH.has(parsed.compute)) {
      if (parsed.architecture === null) {
        throw new LabelValidationError(
          `EC2 compute type '${parsed.compute}' requires architecture label. ` +
          `Must be one of: ${[...EC2_ARCHITECTURES].sort().join(', ')}`
        );
      }
      if (!EC2_ARCHITECTURES.has(parsed.architecture)) {
        throw new LabelValidationError(
          `EC2 platform only supports architectures: ` +
          `${[...EC2_ARCHITECTURES].sort().join(', ')}. Got: ${parsed.architecture}`
        );
      }
    }

    if (EC2_COMPUTE_FORBIDS_ARCH.has(parsed.compute)) {
      if (parsed.architecture !== null) {
        throw new LabelValidationError(
          `EC2 compute type '${parsed.compute}' does not support ` +
          `architecture label. Got: ${parsed.architecture}`
        );
      }
    }
  }
}

function getInstanceType(parsed) {
  if (parsed.platform === 'ecs') {
    return null;
  }
  const key = `${parsed.compute}:${parsed.architecture}`;
  return INSTANCE_TYPE_MAP[key] || null;
}

function getEcsConfig(parsed) {
  if (parsed.platform !== 'ecs') {
    return null;
  }
  if (parsed.compute === 'fargate') {
    return { ...ECS_FARGATE_CONFIG };
  }
  return null;
}

function getTaskArchitecture(parsed) {
  if (parsed.platform !== 'ecs') {
    return null;
  }
  if (parsed.architecture === null) {
    return null;
  }
  return ECS_TASK_ARCHITECTURE_MAP[parsed.architecture] || null;
}

function isSpot(parsed) {
  return parsed.pricing === 'spot';
}

function getRunnerIdNumber(parsed) {
  const match = parsed.runnerId.match(RUNNER_ID_PATTERN);
  if (match) {
    return parseInt(match[1], 10);
  }
  throw new LabelParseError(`Invalid runner ID format: ${parsed.runnerId}`);
}

module.exports = {
  parseLabels,
  validateLabels,
  getInstanceType,
  getEcsConfig,
  getTaskArchitecture,
  isSpot,
  getRunnerIdNumber,
  LabelParseError,
  LabelValidationError,
  // Export constants for testing
  PLATFORMS,
  PRICING_MODELS,
  ECS_COMPUTE,
  ECS_ARCHITECTURES,
  EC2_COMPUTE,
  EC2_ARCHITECTURES,
  COMPUTE_TYPES,
  ALL_ARCHITECTURES
};
